### PR TITLE
Fix gradgrad issues with BatchNorm and Advanced Indexing (v.0.2.0)

### PR DESCRIPTION
**Pull request by gchanan**
This adds two tests for gradchecks:
1) That backward is reentrant
2) That input grad sizes match input sizez.

1) Uncovers an issue with BatchNorm which is now fixed
2) Uncovers an issue with advanced indexing which is also now fixed.

### CODE REVIEW

**Comment by gchanan** [comment]
@killeent for simplicity I fixed the advanced indexing case in python because it's easier, but it should probably be fixed in C++ in case we ever call advanced_indexing_select in other places.  Let's do that in the master version; I'd like to keep this v.0.2.0 version as light as possible.

**Review by soumith — APPROVED** [review]

File: torch/csrc/autograd/functions/batch_normalization.cpp [hunk 1]
```diff
@@ -206,7 +206,7 @@ auto BatchNormBackward::apply(const variable_list& grad_outputs) -> variable_lis
                                  std::move(grad_bias));
   return wrap_outputs(all_inputs, std::move(outputs), [&](FunctionFlags f) {
     return std::make_shared<BatchNormBackwardBackward>(
-      f, *this, std::move(save_mean), std::move(save_std),
```
Comment (soumith) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: just for my own understanding, what makes this non-reentrant, and why is saving references to save_mean / save_std no longer needed?

**Comment by apaszke** [comment]
It's filled in-place when jacobian_c is modified

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -94,21 +94,31 @@ def get_numerical_jacobian(fn, input, target, eps=1e-3):
...
         flat_grad_output.zero_()
         flat_grad_output[i] = 1
-        zero_gradients(input)
-        output.backward(grad_output, retain_graph=True)
-        for jacobian_x, d_x in zip(jacobian, iter_gradients(input)):
-            if d_x is None:
-                jacobian_x[:, i].zero_()
-            else:
-                jacobian_x[:, i] = d_x.to_dense() if d_x.is_sparse else d_x
+        for jacobian_c in (jacobian, jacobian_reentrant):
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: It's filled in-place when jacobian_c is modified

File: torch/csrc/autograd/functions/batch_normalization.cpp (hunk 1 above)
Comment (apaszke) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: It seems that they are not used anywhere in BackwardBackward so there's no point in saving them. A side effect of this move was that save_mean and save_std were set to NULL in Backward, so it segfaulted if you tried to use it again

Comment (gchanan) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: Right.  This makes me realize there are probably two issues here:
1) The double backwards can probably be made more efficient by passing in the saved/running means/vars instead of recomputing
2) We don't actually test BatchNorm single or double backwards with training=False.
